sdk/storage/azure-storage-blob/azure/storage/blob/aio/_upload_helpers.py

```diff
@@ -68,17 +67,27 @@ async def upload_block_blob(  # pylint: disable=too-many-locals, too-many-statem
 
         # Do single put if the size is smaller than config.max_single_put_size
         if adjusted_count is not None and (adjusted_count <= blob_settings.max_single_put_size):
-            try:
-                data = data.read(length)
-                if inspect.isawaitable(data):
-                    data = await data
-                if not isinstance(data, bytes):
-                    raise TypeError('Blob data should be of type bytes.')
-            except AttributeError:
-                pass
+            # Test the output type of the stream and determine if read is awaitable
+            awaitable = False
+            read_type = stream.read(0)
+            if inspect.isawaitable(read_type):
+                awaitable = True
```

Comment:
Logic-wise this seems fine - though I do wonder if there might be a more performant way to achieve this - perhaps we could do a few timeit tests?
Perhaps something like:
```python
try: 
    read_type = await stream.read(0)
except TypeError:
   # Sync stream
else:
   # Async stream
```

Reply:
It was a bit difficult to test (I don't think timeit supports async?) but I did some basic timing tests, and this method was about the same for sync streams and slightly faster with async streams. I think it looks a little cleaner though, so I went for it. Only downside is duplicating the type check/exception.